### PR TITLE
Android O Changes Phase 1

### PR DESCRIPTION
**Pull request by joeygrover**
Android O introduces a lot of new restrictions in terms of background services and when an app becomes activated. This proposal intends to provide enhancements that will work within the new Android O environment, specifically implicit intents and static broadcast receivers.

### CODE REVIEW

**Comment by joeygrover** [comment]
@theresalech This proposal is ready for review

File: proposals/NNNN-android_o_changes_phase_1.md [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+
+## Motivation
+The way the sdl_andorid library currently works will not be functional if the developer targets their app for Android O. There are a few changes we need to make to ensure compatibility, this is the first phase of those changes. The change introduce in the newest version of Android that this proposal is aiming to handle is Android's reduction in accepted implicit intents for static broadcast receivers. An implicit intent is defined by the intent merely having an "action" field, with no specific destination. Previous versions of Android would allow developers to define static broadcast receivers in their manifests that could listen for an intent with that specific "action". The  only implicit intents that developers can use a static broadcast receiver with will be from the Android OS and defined [here](https://developer.android.com/preview/features/background-broadcasts.html). (Note: dynamic broadcast receivers can still receive implicit intents, however, they are subject to the lifecycle of the component in which they are created.)
```
Comment (theresalech) [inline proposals/NNNN-android_o_changes_phase_1.md]: Please update the third sentence to "The change introduced in the newest..."

File: proposals/NNNN-android_o_changes_phase_1.md [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+
+The current SDL Android library uses an implicit intent to broadcast the `SdlRouterService` is currently connected to a module. It is assumed the developers listen for this intent in a static broadcast receiver so the app can be woken up.
```
Comment (theresalech) [inline proposals/NNNN-android_o_changes_phase_1.md]: Can you please clarify this first sentence?  Should it read "...the `SdlRouterService`currently..." (removing "is")?

**Comment by joeygrover** [comment]
All Android O references are with the letter 'O' and not the number '0'.  It only looks that way if you view it in github. Try viewing it [here](https://github.com/joeygrover/sdl_evolution/blob/a36cad8261023937893a3446dbc7506ef65b858f/proposals/NNNN-android_o_changes_phase_1.md).

File: proposals/NNNN-android_o_changes_phase_1.md [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+## Motivation
+The way the sdl_andorid library currently works will not be functional if the developer targets their app for Android O. There are a few changes we need to make to ensure compatibility, this is the first phase of those changes. The change introduced in the newest version of Android that this proposal is aiming to handle is Android's reduction in accepted implicit intents for static broadcast receivers. An implicit intent is defined by the intent merely having an "action" field, with no specific destination. Previous versions of Android would allow developers to define static broadcast receivers in their manifests that could listen for an intent with that specific "action". The  only implicit intents that developers can use a static broadcast receiver with will be from the Android OS and defined [here](https://developer.android.com/preview/features/background-broadcasts.html). (Note: dynamic broadcast receivers can still receive implicit intents, however, they are subject to the lifecycle of the component in which they are created.)
```
Comment (theresalech) [inline proposals/NNNN-android_o_changes_phase_1.md]: Sorry, I missed this in my first review.  In the first sentence, please change to "sdl_android library"

Comment (joeygrover) [inline proposals/NNNN-android_o_changes_phase_1.md]: The project is actually named `sdl_android` so I don't believe this change is necessary. See [here](https://github.com/smartdevicelink/sdl_android).

**Comment by Tnnnguyen** [comment]
We should perform a null check after querying the list of ResolveInfo's since the list can be potentially null or empty as stated [here for queryBroadcastReceivers method](https://developer.android.com/reference/android/content/pm/PackageManager.html#queryBroadcastReceivers(android.content.Intent,int)).
screen shot 2017-07-05 at 11 25 12 am
So the code to send the broadcasts becomes:
```java
List<ResolveInfo> infos = getPackageManager().queryBroadcastReceivers(intent, 0);
if (infos != null) {
	for (ResolveInfo info: infos) {
		startService.setClassName(info.activityInfo.applicationInfo.packageName, info.activityInfo.name);
		sendBroadcast(intent);
	}
}
```

**Comment by joeygrover** [comment]
@Tnnnguyen This proposal is not in review yet. Please save comments for that time. Also, code provided in proposals is not usually 100% what the implementation actually looks like. The proposal is to get an idea of how it will work within the project and to ensure it is a feasible idea. Code completeness should be reviewed during the code review process of the implementation pull request.

**Pull request merged**